src/cdk-experimental/listbox/listbox.ts

```diff
@@ -6,26 +6,112 @@
+import {
+  ContentChildren,
+  Directive,
+  ElementRef,
+  Input,
+  QueryList
+} from '@angular/core';
 
+/**
+ * Directive that applies interaction patterns to an element following the aria role of option.
+ * Typically meant to be placed inside a listbox. Logic handling selection, disabled state, and
+ * value is built in.
+ */
 @Directive({
   selector: '[cdkOption]',
   exportAs: 'cdkOption',
   host: {
     role: 'option',
+    '[attr.aria-selected]': '_selected',
+    '[attr.data-optionid]': '_optionId',
   }
 })
 export class CdkOption {
+  private _selected: boolean | null = null;
+  private _optionId: string;
+
+  /** Whether the option is selected or not */
+  @Input()
+  get selected(): boolean | null {
+    return this._selected;
+  }
+  set selected(value: boolean | null) {
+    this._selected = value;
+  }
+
+  constructor(private readonly _el: ElementRef) {
+    this.setOptionId(`cdk-option-${_uniqueIdCounter++}`);
+  }
 
+  /** Sets the optionId to the given id */
+  setOptionId(id: string): void {
+    this._optionId = id;
+  }
+
+  /** Returns the optionId of this option */
+  getOptionId(): string {
+    return this._optionId;
+  }
+
+  /** Returns an ElementRef of this option */
+  getElementRef(): ElementRef {
+    return this._el;
+  }
 }
 
+let _uniqueIdCounter = 0;
+
+/**
+ * Directive that applies interaction patterns to an element following the aria role of listbox.
+ * Typically CdkOption elements are placed inside the listbox. Logic to handle keyboard navigation,
+ * selection of options, active options, and disabled states is built in.
+ */
 @Directive({
   selector: '[cdkListbox]',
   exportAs: 'cdkListbox',
   host: {
     role: 'listbox',
+    '(click)': 'onClickUpdateSelectedOption($event)'
   }
 })
 export class CdkListbox {
 
+  /** A query list containing all CdkOption elements within this listbox */
+  @ContentChildren(CdkOption, {descendants: true}) _options: QueryList<CdkOption>;
+
+  /** On click handler of this listbox, updates selected value of clicked option */
+  onClickUpdateSelectedOption($event: MouseEvent) {
+    this._options.toArray().forEach(option => {
+      const optionId = option.getOptionId();
+      if ($event.target instanceof Element &&
+          optionId === $event.target?.getAttribute('data-optionid')) {
+        this._updateSelectedOption(option);
+      }
+    });
+  }
+
+  private _updateSelectedOption(option: CdkOption): void {
+    if (option.selected) {
+      this.deselectOption(option);
+    } else {
+      this.selectOption(option);
+    }
+  }
+
+  /** Sets the given option's selected state to true */
+  selectOption(option: CdkOption): void {
+    option.selected = true;
+  }
+
+  /** Sets the given option's selected state to null. Null is preferable for screen readers */
+  deselectOption(option: CdkOption): void {
+    option.selected = null;
+  }
+
+  /** Returns an array of all options that are currently selected */
+  getSelectedOptions(): CdkOption[] {
```

Comment:
I would omit this method for now; I'm not sure it's strictly necessary now that I'm looking at the code. In your tests, you can simply look at the actual options that are selected.

Reply:
Will do. Eventually, do you think this would be a useful public API function for a user to have at hand, even if I don't use it internally?

Reply:
I think we'll have to try building some examples using the component to see; you generally done't want to speculatively add APIs without being totally sure how they wouldn't be used

Reply:
Got it makes sense